### PR TITLE
dataset_purge

### PR DESCRIPTION
**Pull request by davidread**
Adds the `dataset_purge` action.

Issue: #1572

### CODE REVIEW

File: ckan/model/group.py [hunk 1]
```diff
@@ -104,11 +104,11 @@ def related_packages(self):
     def __unicode__(self):
         # refer to objects by name, not ID, to help debugging
         if self.table_name == 'package':
-            table_info = 'package=%s' % meta.Session.query(_package.Package).\
-                get(self.table_id).name
+            pkg = meta.Session.query(_package.Package).get(self.table_id)
+            table_info = 'package=%s' % pkg.name if pkg else 'None'
```
Comment (davidread) [inline ckan/model/group.py]: In the thick of a purge, one of my tests caused an exception here. The Package was purged, yet in the bowels of SQLAlchemy it was doing str(member), so `meta.Session.query(_package.Package).get(self.table_id)` was None and therefore calling `.name` on it caused an Exception. So this just adds an 'if' to cope with that situation.

**Comment by davidread** [comment]
On the issue @wardi noted:
> @mattleduc has done some work on this but ran into a performance issue we'll need to resolve that before adding this to the API.

I've tried this code on DGU. It was taking 1 or 2 minutes to purge each dataset typically, and sometimes for many more minutes. Watching the SQL log, the problem was identifying resource_revisions that it was purging - each took about 1.5 seconds. So I've added a postgres index for resource_revision.continuity_id (in a migration) and now it is down to about typically 10s to purge a dataset and half of that is the COMMIT. I managed to get it to take 21s for a datasets with 2000 resource revisions, and 30s for a monster with 93 resources. There are just lots of SQL calls that SQLAlchemy makes for the purge and cascades etc. I think 10s typical is fine.

**Comment by mattleduc** [comment]
@davidread 10s would be significantly more reasonable, we were up to about 4 minutes for some datasets (with lots of extras and a couple resources)

**Comment by rossjones** [comment]
Tests pass, cli works - on my non-DGU ckan paster took longer to run than the purge.

**Comment by davidread** [comment]
It's worth adding that another thing that slowed purge down in the past was deleting associated user_object_role objects. Migration 78 sorted that out very recently: https://github.com/ckan/ckan/blob/master/ckan/migration/versions/078_remove_old_authz_model.py

**Pull request merged**